File: kirun-js/src/engine/function/system/object/ObjectEntries.ts
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';

import { isNullValue } from '../../../util/NullCheck';
import { EventResult } from '../../../model/EventResult';
import { AbstractObjectFunction } from './AbstractObjectFunction';
import { Schema } from '../../../json/schema/Schema';
import { duplicate } from '../../../util/duplicate';

const VALUE = 'value';

export class ObjectEntries extends AbstractObjectFunction {
    public constructor() {
        super(
            'ObjectEntries',
            Schema.ofArray(
                VALUE,
                Schema.ofArray('tuple', Schema.ofString('key'), Schema.ofAny('value')),
            ),
        );
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        var source = context.getArguments()?.get('source');

        if (isNullValue(source))
            return new FunctionOutput([EventResult.outputOf(new Map([[VALUE, []]]))]);

        let entries = Object.entries(duplicate(source)).sort((a, b) => a[0].localeCompare(b[0]));

        return new FunctionOutput([EventResult.outputOf(new Map([[VALUE, entries]]))]);
    }
}

File: kirun-py/src/kirun_py/function/system/object/object_entries.py
from __future__ import annotations

from typing import TYPE_CHECKING

from kirun_py.function.system.object.abstract_object_function import AbstractObjectFunction
from kirun_py.json.schema.schema import Schema
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput
from kirun_py.util.duplicate import duplicate
from kirun_py.util.null_check import is_null_value

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters

VALUE = 'value'


class ObjectEntries(AbstractObjectFunction):

    def __init__(self) -> None:
        super().__init__(
            'ObjectEntries',
            Schema.of_array(
                VALUE,
                Schema.of_array('tuple', Schema.of_string('key'), Schema.of_any('value')),
            ),
        )

    async def internal_execute(self, context: FunctionExecutionParameters) -> FunctionOutput:
        source = context.get_arguments().get('source') if context.get_arguments() else None

        if is_null_value(source):
            return FunctionOutput([EventResult.output_of({VALUE: []})])

        entries = sorted(duplicate(source).items(), key=lambda x: x[0])
        entries_list = [[k, v] for k, v in entries]

        return FunctionOutput([EventResult.output_of({VALUE: entries_list})])
